crates/ruff_linter/src/rules/flake8_pyi/rules/exit_annotations.rs

```diff
@@ -104,37 +114,52 @@ enum ErrorKind {
     ThirdArgBadAnnotation,
     ArgsAfterFirstFourMustHaveDefault,
     AllKwargsMustHaveDefault,
+    UnrecognizedExitOverload,
 }
 
 /// PYI036
-pub(crate) fn bad_exit_annotation(
-    checker: &mut Checker,
-    is_async: bool,
-    name: &Identifier,
-    parameters: &Parameters,
-) {
+pub(crate) fn bad_exit_annotation(checker: &mut Checker, function: &StmtFunctionDef) {
+    let StmtFunctionDef {
+        is_async,
+        decorator_list,
+        name,
+        parameters,
+        ..
+    } = function;
+
     let func_kind = match name.as_str() {
         "__exit__" if !is_async => FuncKind::Sync,
-        "__aexit__" if is_async => FuncKind::Async,
+        "__aexit__" if *is_async => FuncKind::Async,
         _ => return,
     };
 
-    let positional_args = parameters
-        .args
+    let non_self_positional_args: SmallVec<[&ParameterWithDefault; 3]> = parameters
+        .posonlyargs
         .iter()
-        .chain(parameters.posonlyargs.iter())
-        .collect::<SmallVec<[&ParameterWithDefault; 4]>>();
+        .chain(parameters.args.iter())
+        .skip(1)
+        .collect();
```

Comment:
I think the logic here has been wrong for a while. It should have always been `parameters.posonlyargs.iter().chain(parameters.args.iter())`, not `parameters.args.iter().chain(parameters.posonlyargs.iter())`. I think it didn't come up because it the bug lead to false negatives rather than false positives, so nobody noticed!